Compiler warnings

I document some compiler warnings here:
```
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/xsimd_avx512_conversion.hpp: In function ‘xsimd::batch<double, 8> xsimd::bitwise_cast(const xsimd::batch_bool<double, 8>&)’:
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/../types/xsimd_base.hpp:606:10: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
         *reinterpret_cast<int_type*>(&z1) = ~int_type(0);                      \
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/../types/xsimd_base.hpp:611:5: note: in expansion of macro ‘XSIMD_DEFINE_BITWISE_CAST_FLOAT’
     XSIMD_DEFINE_BITWISE_CAST_FLOAT(double, NMIN)                              \
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5: note: in expansion of macro ‘XSIMD_DEFINE_BITWISE_CAST_ALL’
     XSIMD_DEFINE_BITWISE_CAST_ALL(8)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/xsimd_avx512_conversion.hpp: In function ‘xsimd::batch<float, 16> xsimd::bitwise_cast(const xsimd::batch_bool<float, 16>&)’:
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/../types/xsimd_base.hpp:606:10: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
         *reinterpret_cast<int_type*>(&z1) = ~int_type(0);                      \
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/../types/xsimd_base.hpp:612:5: note: in expansion of macro ‘XSIMD_DEFINE_BITWISE_CAST_FLOAT’
     XSIMD_DEFINE_BITWISE_CAST_FLOAT(float, NMIN * 2)                           \
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tdegeus/miniconda3/envs/code_velocity_s6g1/include/xsimd/types/xsimd_avx512_conversion.hpp:47:5: note: in expansion of macro ‘XSIMD_DEFINE_BITWISE_CAST_ALL’
     XSIMD_DEFINE_BITWISE_CAST_ALL(8)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/.azure-pipelines/unix-build.yml
+++ b/.azure-pipelines/unix-build.yml
@@ -39,7 +39,7 @@ steps:
         CMAKE_EXTRA_ARGS="$CMAKE_EXTRA_ARGS -DXSIMD_FORCE_X86_INSTR_SET=0 -DXSIMD_FORCE_X86_AMD_INSTR_SET=0";
       fi
 
-      cmake -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX $CMAKE_EXTRA_ARGS -DDOWNLOAD_GTEST=ON $(Build.SourcesDirectory)
+      cmake -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX $CMAKE_EXTRA_ARGS -DDOWNLOAD_GTEST=ON -DXSIMD_ENABLE_WERROR=1 $(Build.SourcesDirectory)
     displayName: Configure xsimd
     workingDirectory: $(Build.BinariesDirectory)
 

--- a/include/xsimd/config/xsimd_instruction_set.hpp
+++ b/include/xsimd/config/xsimd_instruction_set.hpp
@@ -52,34 +52,42 @@
 #ifdef XSIMD_FORCE_X86_INSTR_SET
     #define XSIMD_X86_INSTR_SET XSIMD_FORCE_X86_INSTR_SET
     #define XSIMD_X86_INSTR_SET_AVAILABLE XSIMD_VERSION_NUMBER_AVAILABLE
-    #ifdef _MSC_VER
-        #pragma message("Warning: Forcing X86 instruction set")
-    #else
-        #warning "Forcing X86 instruction set"
+    #ifndef XSIMD_SKIP_ON_WERROR
+        #ifdef _MSC_VER
+            #pragma message("Warning: Forcing X86 instruction set")
+        #else
+            #warning "Forcing X86 instruction set"
+        #endif
     #endif
 #elif defined(XSIMD_FORCE_X86_AMD_INSTR_SET)
     #define XSIMD_X86_AMD_INSTR_SET XSIMD_FORCE_X86_AMD_INSTR_SET
     #define XSIMD_X86_AMD_INSTR_SET_AVAILABLE XSIMD_VERSION_NUMBER_AVAILABLE
-    #ifdef _MSC_VER
-        #pragma message("Warning: Forcing X86 AMD instruction set")
-    #else
-        #warning "Forcing X86 AMD instruction set"
+    #ifndef XSIMD_SKIP_ON_WERROR
+        #ifdef _MSC_VER
+            #pragma message("Warning: Forcing X86 AMD instruction set")
+        #else
+            #warning "Forcing X86 AMD instruction set"
+        #endif
     #endif
 #elif defined(XSIMD_FORCE_PPC_INSTR_SET)
     #define XSIMD_PPC_INSTR_SET XSIMD_FORCE_PPC_INSTR_SET
     #define XSIMD_PPC_INSTR_SET_AVAILABLE XSIMD_VERSION_NUMBER_AVAILABLE
-    #ifdef _MSC_VER
-        #pragma message("Warning: Forcing PPC instruction set")
-    #else
-        #warning "Forcing PPC instruction set"
+    #ifndef XSIMD_SKIP_ON_WERROR
+        #ifdef _MSC_VER
+            #pragma message("Warning: Forcing PPC instruction set")
+        #else
+            #warning "Forcing PPC instruction set"
+        #endif
     #endif
 #elif defined(XSIMD_FORCE_ARM_INSTR_SET)
     #define XSIMD_ARM_INSTR_SET XSIMD_FORCE_ARM_INSTR_SET
     #define XSIMD_ARM_INSTR_SET_AVAILABLE XSIMD_VERSION_NUMBER_AVAILABLE
-    #ifdef _MSC_VER
-        #pragma message("Warning: Forcing ARM instruction set")
-    #else
-        #warning "Forcing ARM instruction set"
+    #ifndef XSIMD_SKIP_ON_WERROR
+        #ifdef _MSC_VER
+            #pragma message("Warning: Forcing ARM instruction set")
+        #else
+            #warning "Forcing ARM instruction set"
+        #endif
     #endif
 #endif
 

--- a/include/xsimd/types/xsimd_avx_float.hpp
+++ b/include/xsimd/types/xsimd_avx_float.hpp
@@ -707,7 +707,7 @@ namespace xsimd
             }
 
             template<bool... Values>
-            static batch_type select(const batch_bool_constant<value_type, Values...>& cond, const batch_type& a, const batch_type& b)
+            static batch_type select(const batch_bool_constant<value_type, Values...>&, const batch_type& a, const batch_type& b)
             {
                 constexpr int mask = batch_bool_constant<value_type, Values...>::mask();
                 return _mm256_blend_ps(b, a, mask);

--- a/include/xsimd/types/xsimd_base_constant.hpp
+++ b/include/xsimd/types/xsimd_base_constant.hpp
@@ -29,7 +29,7 @@ namespace xsimd
 
         bool operator[](size_t i) const
         {
-            return std::array<value_type, size>{Values...}[i];
+            return std::array<value_type, size>{{Values...}}[i];
         }
         static constexpr int mask()
         {

--- a/include/xsimd/types/xsimd_fallback.hpp
+++ b/include/xsimd/types/xsimd_fallback.hpp
@@ -603,7 +603,8 @@ namespace xsimd
             {
                 T res(0);
                 using int_type = as_unsigned_integer_t<T>;
-                *reinterpret_cast<int_type*>(&res) = ~int_type(0);
+                int_type value(~int_type(0));
+                std::memcpy(&res, &value, sizeof(int_type));
                 return res;
             }
         };

--- a/include/xsimd/types/xsimd_sse_double.hpp
+++ b/include/xsimd/types/xsimd_sse_double.hpp
@@ -633,6 +633,7 @@ namespace xsimd
             static batch_type select(const batch_bool_constant<value_type, Values...>& cond, const batch_type& a, const batch_type& b)
             {
 #if XSIMD_X86_INSTR_SET >= XSIMD_X86_SSE4_1_VERSION
+                (void)cond;
                 constexpr int mask = batch_bool_constant<value_type, Values...>::mask();
                 return _mm_blend_pd(b, a, mask);
 #else

--- a/include/xsimd/types/xsimd_types_include.hpp
+++ b/include/xsimd/types/xsimd_types_include.hpp
@@ -70,18 +70,21 @@
 #endif
 
 #if !defined(XSIMD_INSTR_SET_AVAILABLE)
-    #if defined(XSIMD_ENABLE_FALLBACK)
-        #ifdef _MSC_VER
-	    #pragma message("Warning: No SIMD instructions set detected, using fallback mode.")
+    #ifndef XSIMD_SKIP_ON_WERROR
+
+        #if defined(XSIMD_ENABLE_FALLBACK)
+            #ifdef _MSC_VER
+	              #pragma message("Warning: No SIMD instructions set detected, using fallback mode.")
+            #else
+	              #warning "No SIMD instructions set detected, using fallback mode."
+            #endif
         #else
-	    #warning "No SIMD instructions set detected, using fallback mode."
-        #endif
-    #else
-        #ifdef _MSC_VER
-	    #pragma message("Warning: No SIMD instructions set detected, please enable SIMD instructions or activate the fallback mode. (e.g. for x86 -march=native or for ARM -mfpu=neon)")
-        #else
-	    #warning "No SIMD instructions set detected, please enable SIMD instructions or activate the fallback mode. (e.g. for x86 -march=native or for ARM -mfpu=neon)"
-	#endif
+            #ifdef _MSC_VER
+	              #pragma message("Warning: No SIMD instructions set detected, please enable SIMD instructions or activate the fallback mode. (e.g. for x86 -march=native or for ARM -mfpu=neon)")
+            #else
+	              #warning "No SIMD instructions set detected, please enable SIMD instructions or activate the fallback mode. (e.g. for x86 -march=native or for ARM -mfpu=neon)"
+	          #endif
+	      #endif
     #endif
 #endif
 

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -30,6 +30,8 @@ include(CheckCXXCompilerFlag)
 
 string(TOUPPER "${CMAKE_BUILD_TYPE}" U_CMAKE_BUILD_TYPE)
 
+OPTION(XSIMD_ENABLE_WERROR "Turn on -Werror" OFF)
+
 ################
 # ARM SETTINGS #
 ################
@@ -238,4 +240,8 @@ if (CROSS_COMPILE_ARM)
     add_custom_target(xtest COMMAND qemu-arm -L /usr/arm-linux-gnueabi/ test_xsimd DEPENDS test_xsimd)
 else()
     add_custom_target(xtest COMMAND test_xsimd DEPENDS test_xsimd)
+endif()
+
+if (XSIMD_ENABLE_WERROR)
+    target_compile_options(test_xsimd PRIVATE -Werror -Wall -DXSIMD_SKIP_ON_WERROR)
 endif()

--- a/test/test_algorithms.cpp
+++ b/test/test_algorithms.cpp
@@ -263,6 +263,7 @@ TEST(algorithms, iterator)
     *cbegin = sin(*cbegin);
     *cbegin = sqrt(*cbegin);
     auto real_part = abs(*(cbegin));
+    (void)real_part;
 #endif
 
 }

--- a/test/test_basic_math.cpp
+++ b/test/test_basic_math.cpp
@@ -98,8 +98,8 @@ protected:
         }
         // clip
         {
-            value_type clip_lo = 0.5;
-            value_type clip_hi = 1.;
+            value_type clip_lo = static_cast<value_type>(0.5);
+            value_type clip_hi = static_cast<value_type>(1.);
             array_type expected;
             std::transform(clip_input.cbegin(), clip_input.cend(), expected.begin(),
                            [clip_lo, clip_hi](const value_type& l) {

--- a/test/test_batch_int.cpp
+++ b/test/test_batch_int.cpp
@@ -32,7 +32,7 @@ namespace xsimd
 
             T max = std::numeric_limits<T>::max();
             T min = std::numeric_limits<T>::min();
-            std::array<T, 2> maxmin_cmp{max, min};
+            std::array<T, 2> maxmin_cmp{{max, min}};
             B maxmin(max, min);
             EXPECT_BATCH_EQ(maxmin, maxmin_cmp) << print_function_name("numeric max and min");
 
@@ -43,8 +43,8 @@ namespace xsimd
             auto r1 = xsimd::max(a, c);
             auto r3 = xsimd::min(a, c);
 
-            EXPECT_BATCH_EQ(r1, A({2, 3})) << print_function_name("max");
-            EXPECT_BATCH_EQ(r3, A({1, 3})) << print_function_name("min");
+            EXPECT_BATCH_EQ(r1, (A{{2, 3}})) << print_function_name("max");
+            EXPECT_BATCH_EQ(r3, (A{{1, 3}})) << print_function_name("min");
 
             auto r4 = a < b; // test lt
             BB e4(1, 0);
@@ -68,8 +68,8 @@ namespace xsimd
             auto r1 = xsimd::max(a, c);
             auto r3 = xsimd::min(a, c);
 
-            EXPECT_BATCH_EQ(r1, A({2, 3, 2, 3})) << print_function_name("max");
-            EXPECT_BATCH_EQ(r3, A({1, 3, 1, 1})) << print_function_name("min");
+            EXPECT_BATCH_EQ(r1, (A{{2, 3, 2, 3}})) << print_function_name("max");
+            EXPECT_BATCH_EQ(r3, (A{{1, 3, 1, 1}})) << print_function_name("min");
 
             auto r4 = a < b; // test lt
             BB e4(1,0,1,1);
@@ -88,7 +88,7 @@ namespace xsimd
 
             T max = std::numeric_limits<T>::max();
             T min = std::numeric_limits<T>::min();
-            std::array<T, 8> maxmin_cmp{0, 0, max, 0, min, 0, 0, 0};
+            std::array<T, 8> maxmin_cmp{{0, 0, max, 0, min, 0, 0, 0}};
             B maxmin(0, 0, max, 0, min, 0, 0, 0);
             EXPECT_BATCH_EQ(maxmin, maxmin_cmp) << print_function_name("numeric max and min");
 
@@ -99,8 +99,8 @@ namespace xsimd
             auto r1 = xsimd::max(a, c);
             auto r3 = xsimd::min(a, c);
             auto r4 = a < b; // test lt
-            EXPECT_BATCH_EQ(r1, A({2, 3, 2, 3, 2, 3, 3, 3})) << print_function_name("max");
-            EXPECT_BATCH_EQ(r3, A({1, 3, 1, 3, 1, 1, 2, 3})) << print_function_name("min");
+            EXPECT_BATCH_EQ(r1, (A{{2, 3, 2, 3, 2, 3, 3, 3}})) << print_function_name("max");
+            EXPECT_BATCH_EQ(r3, (A{{1, 3, 1, 3, 1, 1, 2, 3}})) << print_function_name("min");
 
             BB e4(1,0,1,0, 1,1,0,0);
             EXPECT_TRUE(xsimd::all(r4 == e4));
@@ -118,7 +118,7 @@ namespace xsimd
 
             T max = std::numeric_limits<T>::max();
             T min = std::numeric_limits<T>::min();
-            std::array<T, 16> maxmin_cmp{0, 0, max, 0, min, 0, 0, 0, 0, 0, max, 0, min, 0, 0, 0};
+            std::array<T, 16> maxmin_cmp{{0, 0, max, 0, min, 0, 0, 0, 0, 0, max, 0, min, 0, 0, 0}};
             B maxmin(0, 0, max, 0, min, 0, 0, 0, 0, 0, max, 0, min, 0, 0, 0);
             EXPECT_BATCH_EQ(maxmin, maxmin_cmp) << print_function_name("numeric max and min");
 
@@ -131,8 +131,8 @@ namespace xsimd
             auto r5 = a == c;
             auto r6 = a != c;
 
-            EXPECT_BATCH_EQ(r1, A({2,3,2,3, 2,3,2,3, 3,3,3,3, 2,max,max,2})) << print_function_name("max");
-            EXPECT_BATCH_EQ(r3, A({1,2,1,2, 1,2,1,2, 2,2,2,2, min,2,2,min})) << print_function_name("min");
+            EXPECT_BATCH_EQ(r1, (A{{2,3,2,3, 2,3,2,3, 3,3,3,3, 2,max,max,2}})) << print_function_name("max");
+            EXPECT_BATCH_EQ(r3, (A{{1,2,1,2, 1,2,1,2, 2,2,2,2, min,2,2,min}})) << print_function_name("min");
 
             BB e4(1,0,1,0, 1,0,1,0, 0,0,0,0, 1,0,0,1);
             EXPECT_TRUE(xsimd::all(r4 == e4));
@@ -161,8 +161,8 @@ namespace xsimd
             auto r1 = xsimd::max(a, b);
             auto r3 = xsimd::min(a, b);
             auto r4 = a < b; // test lt
-            EXPECT_BATCH_EQ(r1, A({2,3,2,3, 2,3,2,3, 2,3,2,3, 2,3,2,3, 2,3,2,3, 2,3,2,3, 3,3,3,3, 2,max,max,2})) << print_function_name("max");
-            EXPECT_BATCH_EQ(r3, A({1,2,1,2, 1,2,1,2, 1,2,1,2, 1,2,1,2, 1,2,1,2, 1,2,1,2, 2,2,2,2, min,2,2,min})) << print_function_name("min");
+            EXPECT_BATCH_EQ(r1, (A{{2,3,2,3, 2,3,2,3, 2,3,2,3, 2,3,2,3, 2,3,2,3, 2,3,2,3, 3,3,3,3, 2,max,max,2}})) << print_function_name("max");
+            EXPECT_BATCH_EQ(r3, (A{{1,2,1,2, 1,2,1,2, 1,2,1,2, 1,2,1,2, 1,2,1,2, 1,2,1,2, 2,2,2,2, min,2,2,min}})) << print_function_name("min");
 
             BB e4(1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 1,0,1,0, 0,0,0,0, 1,0,0,1);
             EXPECT_TRUE(xsimd::all(r4 == e4));

--- a/test/test_utils.hpp
+++ b/test/test_utils.hpp
@@ -133,28 +133,36 @@ namespace detail
         }
     }
 
+    inline unsigned char uabs(unsigned char val)
+    {
+        return val;
+    }
+
+    inline unsigned short uabs(unsigned short val)
+    {
+        return val;
+    }
+
+    inline unsigned int uabs(unsigned int val)
+    {
+        return val;
+    }
+
+    inline unsigned long uabs(unsigned long val)
+    {
+        return val;
+    }
+
+    inline unsigned long long uabs(unsigned long long val)
+    {
+        return val;
+    }
+
     template <class T>
     inline T uabs(T val)
     {
         return std::abs(val);
     }
-
-    inline uint32_t uabs(uint32_t val)
-    {
-        return val;
-    }
-
-    inline uint64_t uabs(uint64_t val)
-    {
-        return val;
-    }
-
-#ifdef XSIMD_32_BIT_ABI
-    inline unsigned long uabs(unsigned long val)
-    {
-        return val;
-    }
-#endif
 
     template <class T>
     bool check_is_small(const T& value, const T& tolerance)
@@ -170,9 +178,9 @@ namespace detail
         {
             return (std::numeric_limits<T>::max)();
         }
-        if (lhs == static_cast<T>(0) ||
-            rhs > static_cast<T>(1) &&
-                lhs < rhs * (std::numeric_limits<T>::min)())
+        if ((lhs == static_cast<T>(0)) ||
+            (rhs > static_cast<T>(1) &&
+                lhs < rhs * (std::numeric_limits<T>::min)()))
         {
             return static_cast<T>(0);
         }

